sirepo/package_data/static/js/sirepo.js

```diff
@@ -1556,14 +1544,154 @@ SIREPO.app.service('validationService', function(utilities) {
 
 });
 
+SIREPO.app.factory('srCache', function(appState, $rootScope) {
+
+    // Browser side caching implemented using indexedDB
+    // - Caches sim frame responses
+    // - Allows cache to be cleared by simId or (simId, modelName)
+    // - Keeps MRU date on item access and deletes expired records at startup
+
+    const self = {};
+    const STORE = 'db';
+    const FRAME = 'frame';
+    // 30 days until expired
+    const EXPIRY_TIME = 30 * 24 * 60 * 60 * 1000;
+    let db = null;
+
+    function deleteKeys(keys) {
+        if (keys.length) {
+            const o = objectStore('readwrite');
+            if (o) {
+                for (const k of keys) {
+                    o.delete(k);
+                }
+            }
+        }
+    }
+
+    function invokeCallback(callback, value) {
+        $rootScope.$applyAsync(() => callback(value));
+    }
+
+    function key(prefix, value) {
+        return prefix + ':' + value;
+    }
+
+    function objectStore(mode) {
+        try {
+            if (db) {
```

Comment:
I think there's a race condition here on startup. It may be that db is possible, but onsuccess hasn't been called yet.

Reply:
Yes, it could be that onsuccess() call is pending, but a call to getFrame would return a cache-miss. I think that is better than having to wait for the browser to sort out the database if there is a problem. In practice the db seems to initialize very quickly.

Reply:
@robnagler I've made some improvements with 497424e